# Game Of Learners Clinic
A Tech Interview Prep program to help students from all walks of life to level up on
data structures, algorithms, system design and other Computer Science Fundamentals.
## Entry Challenge v.2
Write a function CountZeros(A) that takes in an array of integers A, and returns the number of 0's in that array. For example, given [1, 0, 5, 6, 0, 2], the function/method should return 2.

## Solution
To tackle this challenge we shall attempt different approaches to find an optimum by execution speed and code length

Imports

In [1]:
# The below imports are Python Libraries that will enable use to manipulate data

# Pandas is mainly used for data analysis
import pandas as pd
# NumPy is a Python library used for working with arrays
import numpy as np
# Timeit is a Python library to measure the execution time taken by the given code snippet
import timeit 

Array

In [2]:
# Lets create arrays and name it them: These are the arrays we shall use to test our functions

A = np.array([1, 0, 5, 6, 0, 2]) # 0D Array
B = np.array([[[1, 0, 3], [4, 5, 0]], [[1, 2, 3], [0, 5, 6]]]) # A multi-dimensional array(3D)

# Let's Call out the arrays we've created 
print("A: \n",A,'\n'*2,"B: \n", B)


A: 
 [1 0 5 6 0 2] 

 B: 
 [[[1 0 3]
  [4 5 0]]

 [[1 2 3]
  [0 5 6]]]


### Approach 1:
In this approach we shall directly count the number of 0's in the array using a Logic operator

In [3]:
# This code uses the Logic operator (==) to check where the integer in an array(A) is zero, then returns the total for such occurences
print(np.count_nonzero(A==0))

2


In [4]:
# We shall then create a function named CountZeros then fit the code in it:

def CountZeros(A):          # Where A represents the name of the array to check
    number_of_zeros = np.count_nonzero(A == 0)
    return number_of_zeros  # Output after counting zeros in array

print("Number of 0's: ",CountZeros(B))

# Let's Check how fast this function is:
%timeit CountZeros(B)

Number of 0's:  3
The slowest run took 19.13 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.81 µs per loop


### Approach 2:
In this approach we shall subtract the number of integers that are not 0's from the size of the array to get the number of 0's.

In [5]:
# We shall create the function then test it out on array B since it is multidimensional

def CountZeros(A):        # Where A represents the name of the array to check
  number_of_zeros = A.size - np.count_nonzero(A) # Where we subtract non-zeros from lenth of array
  return number_of_zeros  # Output after counting 0's in array

print("Number of 0's:",CountZeros(B))

# Let's Check how fast this function is:
%timeit CountZeros(B)

Number of 0's: 3
The slowest run took 13.25 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 847 ns per loop


## Results
- Approach 1 executed in 1.81 µs per loop (1.81 * 10*-6 Seconds)
- Approach 2 executed in 847 ns per loop (847 * 10*-9 Seconds)


Approach 2 is 0.48 times faster than Approach 1 even though it is the indirect approach.

Both approaches work on multi-dimensional and zero dimensional arrays.